Support for local (private?) IGV tracks

northwestwitch opened the pull request:
### DO NOT BE SCARED BY THE AMOUNT OF FILES WITH CHANGES; IT'S MOSTLY RENAMING OF VARIABLES :)

This PR adds a functionality or fixes a bug.
- Closes #4509 
- This will open up to using private tracks that are not publicly available to download, i.e. Cosmic
- Can be tracks of any sort, public or private, the important is that they are respecting the standards of igv.js

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>


**How to test**:
1. Uncomment the IGV tracks settings on the demo instance of Scout.
2. Go to any MT variant of the demo case 643594
3. Move to a gene in another chromosome (perhaps a known gene associated to diseases) and make sure that the MANE track is visible

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by DN
- [x] tests executed by CR

northwestwitch commented on scout/server/app.py:
```diff
@@ -158,9 +159,9 @@ def configure_extensions(app):
-    if app.config.get("CLOUD_IGV_TRACKS"):
-        LOG.info("Collecting IGV tracks from cloud resources")
-        extensions.cloud_tracks.init_app(app)
+    if app.config.get("CUSTOM_IGV_TRACKS") or app.config.get("CLOUD_IGV_TRACKS"):
```
Renamed the variable in the config file, but with backward compatibility

northwestwitch commented on scout/utils/track_resources.py:
```diff
@@ -0,0 +1,75 @@
+# coding=UTF-8
```
This is simply the file named cloud_resources.py which has been renamed

northwestwitch replied:
Never heard of it 🤣

northwestwitch commented on scout/utils/track_resources.py:
```diff
@@ -0,0 +1,75 @@
+        if all(track.get(key) for key in REQUIRED_FIELDS):
+            return track
+
+    def set_local_track_path(self, path: str) -> str:
```
This is a new function

northwestwitch commented on scout/server/blueprints/alignviewers/templates/alignviewers/igv_viewer.html:
```diff
@@ -64,13 +64,22 @@
                         },
                       {% endfor %}
-                      {% for custTrack in cloud_public_tracks %}
+                      {% for custTrack in config_custom_tracks %}
```
Renamed cloud_public_tracks -> **config**_custom_tracks, because it's not obvious that the tracks are in the cloud and it's easier to understand that they are specified in the config file

northwestwitch commented:
Oh no, something broke the test in this one, I'll fix it, but not today

dnil reviewed (commented):
Since we are on a review spree, I might comment on this one as well. I tested locally, and noted that the local `mane.bb` track does not display (tried both builds). The remote example is still working though! There are also a few residual issues with how the custom tracks get added to the user igv_tracks, especially if the user tries to silence them by unsetting all, the custom tracks get added in. They are probably already there with main, but going unnoticed since we dont use the feature much. I guess I will revisit later, but can you confirm that you actually see the mane track locally? 😊

northwestwitch commented on scout/server/blueprints/alignviewers/views.py:
```diff
@@ -77,10 +76,6 @@ def remote_static():
     if controllers.check_session_tracks(file_path) is False:
         return abort(403)
 
-    range_header = request.headers.get("Range", None)
```
Just the check, not the use of ranged request. That's to allow ALSO unranged request to use this endpoint, as in this functionality.

dnil replied:
Yes, but it only required for bam/cram files?

northwestwitch replied:
> Yes, but it only required for bam/cram files?

Perhaps required also for some tracks, I don't remember right now

dnil replied:
😆What I mean is look at the next couple
of lines: this check would return an error only for bam and cram before, not at all now. I don’t see why?

northwestwitch replied:
We are sending ranged queries for bam/cram anyway, so it should be fine

northwestwitch replied:
But I'll check again why I removed it in this PR

northwestwitch replied:
I think I've removed these lines while trying to understand why I couldn't see the new track. Having found the issue somewhere else I forgot to restore this code. Now it's back again. Good catch!

northwestwitch commented:
I would keep this PR outside the upcoming release, so we'll have more time to test it

northwestwitch commented:
I've fixed the things that you've pointed out, I think:

>  I tested locally, and noted that the local `mane.bb` track does not display (tried both builds). 

I've now defined a height for the custom tracks and shows in my local instance (I think this was the problem). In order to see the track you should select a MT variant.

> a few residual issues with how the custom tracks get added to the user igv_tracks, especially if the user tries to silence them by unsetting all, the custom tracks get added in.

Right, I've introduced a fix to this select:

<img width="803" alt="image" src="https://github.com/Clinical-Genomics/scout/assets/28093618/9065cb2f-9c26-4cd7-a9e8-7a36676300ff">

The problem was [here](https://github.com/Clinical-Genomics/scout/blob/d448461287114e5a79d684ed2e67c913aea1906e/scout/server/blueprints/variant/controllers.py#L375): even from a MT variant, the genome build passed to get_igv_tracks was always "37". This shouls be fixed in this branch. Note for variants not on chr MT you won't see the MANE track option, unless the case has build 38. But you'll see the option on MT variants for cases with build 37.

northwestwitch commented on scout/server/config.py:
```diff
@@ -99,6 +98,19 @@
+#        "access": "public",
+#        "tracks": [
+#            {
+#                "name": "NCBI RefSeq & Ensembl transcripts from MANE",
```
This name is super long, perhaps we could call the track something shorter

dnil replied:
Yes, "MANE" comes to mind.. 😸

dnil commented on scout/utils/track_resources.py:
```diff
@@ -0,0 +1,75 @@
+                    )
+                    continue
+
+                if bool(re.match("https?://", track_obj["url"])) is False:  # Is a local file
```
Ping on this. 😉

dnil commented on tests/utils/test_custom_tracks.py:
```diff
@@ -0,0 +1,55 @@
+    assert config_igv_tracks.tracks[TRACK_BUILD][0]["url"] == TRACK_URL
+
+
+def test_align_handler_custom_local_tracks():
```
👍

northwestwitch replied:
Fixed the re.compile as well. Thanks for the new review, merging now!

dnil reviewed (approved):
Tentatively approving since it's working just fine now - thank you! But please see the comments about return of track version (will break stuff for the native "38" users as it is) and preferably the RE performance enhancement through compilation.

northwestwitch commented on scout/server/blueprints/variant/controllers.py:
```diff
@@ -372,7 +372,7 @@ def variant(
         "case_tag_options": CASE_TAGS,
         "inherit_palette": INHERITANCE_PALETTE,
-        "igv_tracks": get_igv_tracks(genome_build),
+        "igv_tracks": get_igv_tracks("38" if variant_obj["is_mitochondrial"] else "37"),
```
Mmm right, of course I tried it with a 37 case, thanks!
